In [1]:
import numpy as np
import pandas as pd
import sklearn
import matplotlib.pyplot as plt

### PLAN

1. Pobranie danych
2. Przejrzeć co mamy
    - wielkosc data setu
    - jakie kolumny
    - jakiego typu dane
    - podejrzec probke
    - jesli dane sa we wlasciwym formacie: proste statystyki
3. Konwersja danych na odpowiedni format
4. Co z brakujacymi wartosciami?

#### Zaimportuj i przejrzyj dane

In [2]:
dane = pd.read_csv('data/DataWorkshop.csv')

In [3]:
print('Rozmiar danych: ', dane.shape)

Rozmiar danych:  (229929, 30)


In [4]:
print('Kolumny w datasecie: ')
dane.columns

Kolumny w datasecie: 


Index(['Data zdarzenia', 'Godzina', 'GPS x', 'GPS y', 'Miejscowość', 'Ulica',
       'Numer domu', 'Numer drogi', 'KM HM (Pikietaż)', 'Obszar',
       'Skrzyżowanie', 'Geometria drogi', 'Charakterystyka miejsca zdarzenia',
       'Oświetlenie', 'Warunki atmosferyczne', 'Rodzaj zdarzenia', 'Wina',
       'Rodzaj drogi', 'Nawierzchnia drogi', 'Stan nawierzchni',
       'Sygnalizacja świetlna', 'Oznakowanie poziome',
       'Liczba pasów w zdarzeniu', 'Zachowanie kierującego',
       'Zachowanie pieszego', 'Inne przyczyny', 'Liczba rannych w zdarzeniu',
       'Liczba zabitych w zdarzeniu', 'Liczba kolizji', 'Liczba wypadków'],
      dtype='object')

In [5]:
print('Typy danych: ')
dane.dtypes

Typy danych: 


Data zdarzenia                        object
Godzina                               object
GPS x                                 object
GPS y                                 object
Miejscowość                           object
Ulica                                 object
Numer domu                            object
Numer drogi                           object
KM HM (Pikietaż)                      object
Obszar                                object
Skrzyżowanie                          object
Geometria drogi                       object
Charakterystyka miejsca zdarzenia     object
Oświetlenie                           object
Warunki atmosferyczne                 object
Rodzaj zdarzenia                      object
Wina                                  object
Rodzaj drogi                          object
Nawierzchnia drogi                    object
Stan nawierzchni                      object
Sygnalizacja świetlna                 object
Oznakowanie poziome                   object
Liczba pas

In [6]:
print('Brakujace wartosci: ')

missing = dane.isnull().sum()
print(missing)

Brakujace wartosci: 
Data zdarzenia                            0
Godzina                                   0
GPS x                                     0
GPS y                                     0
Miejscowość                               0
Ulica                                 20728
Numer domu                            93884
Numer drogi                               0
KM HM (Pikietaż)                     194888
Obszar                                    1
Skrzyżowanie                         166779
Geometria drogi                       66008
Charakterystyka miejsca zdarzenia         2
Oświetlenie                               0
Warunki atmosferyczne                     1
Rodzaj zdarzenia                          0
Wina                                      0
Rodzaj drogi                              1
Nawierzchnia drogi                        1
Stan nawierzchni                          1
Sygnalizacja świetlna                     0
Oznakowanie poziome                       2
Liczba pasó

In [7]:
print('Brakujace wartosci jako procent zbioru, %: ')
index=[]
lst=[]
for item in missing.iteritems():
    index.append(item[0])
    lst.append(round((item[1]/dane.shape[0])*100,4))
missing_percentage=pd.Series(lst, index=index)
print(missing_percentage)

Brakujace wartosci jako procent zbioru, %: 
Data zdarzenia                        0.0000
Godzina                               0.0000
GPS x                                 0.0000
GPS y                                 0.0000
Miejscowość                           0.0000
Ulica                                 9.0150
Numer domu                           40.8317
Numer drogi                           0.0000
KM HM (Pikietaż)                     84.7601
Obszar                                0.0004
Skrzyżowanie                         72.5350
Geometria drogi                      28.7080
Charakterystyka miejsca zdarzenia     0.0009
Oświetlenie                           0.0000
Warunki atmosferyczne                 0.0004
Rodzaj zdarzenia                      0.0000
Wina                                  0.0000
Rodzaj drogi                          0.0004
Nawierzchnia drogi                    0.0004
Stan nawierzchni                      0.0004
Sygnalizacja świetlna                 0.0000
Oznakowanie

In [8]:
# Probka danych
dane.sample(50)

Data zdarzenia   Godzina      GPS x      GPS y           Miejscowość  \
173098     2019-01-27  17:00:00  19*07'199  50*48'408           CZĘSTOCHOWA   
44027      2016-11-06  16:30:00  18*49'258  50*26'404       TARNOWSKIE GÓRY   
54028      2017-01-07  11:00:00  19*00'288  50*15'446              KATOWICE   
212486     2019-09-27  00:35:00  19*26'044  50*28'569             ZAWIERCIE   
193476     2019-05-31  16:13:00  18*50'208  50*08'324         ŁAZISKA GÓRNE   
132421     2018-05-23  12:30:00  19*04'190  49*53'198              JANOWICE   
8063       2016-03-03  13:45:00  18*54'491  50*17'088        ŚWIĘTOCHŁOWICE   
143977     2018-08-08  08:41:00  19*15'245  50*20'054      DĄBROWA GÓRNICZA   
192434     2019-05-25  11:00:00  18*54'121  50*15'125           RUDA ŚLĄSKA   
11105      2016-03-27  22:20:00  19*23'589  50*27'275     KUŹNICA MASŁOŃSKA   
183704     2019-03-31  15:14:00  19*00'393  50*06'471                 TYCHY   
26631      2016-07-14  21:40:00  19*13'089  50*19'558      DĄBROWA GÓRNICZA   
37021      2016-09-26  18:00:00  18*37'523  50*23'553             PYSKOWICE   
81587      2017-07-11  20:10:00  18*57'407  50*16'291              KATOWICE   
188084     2019-04-29  12:40:00  19*03'398  49*48'067         BIELSKO-BIAŁA   
157610     2018-10-28  13:10:00  18*47'450  50*17'292                ZABRZE   
161391     2018-11-21  16:00:00  19*03'461  50*16'335              KATOWICE   
209911     2019-09-10  07:25:00  18*45'070  50*18'552                ZABRZE   
124674     2018-04-03  19:50:00  18*38'144  50*18'140               GLIWICE   
200370     2019-07-12  10:02:00  19*23'320  50*29'071             ZAWIERCIE   
180802     2019-03-15  15:40:00  19*04'174  50*13'460              KATOWICE   
13478      2016-04-14  18:20:00  19*02'480  50*14'350              KATOWICE   
144483     2018-08-11  00:48:00  18*54'064  50*54'448               KŁOBUCK   
27557      2016-07-21  15:30:00  19*00'429  50*18'492  SIEMIANOWICE ŚLĄSKIE   
41177      2016-10-20  14:20:00  18*40'589  50*18'157               GLIWICE   
118202     2018-02-19  07:11:10  19*06'415  50*47'497           CZĘSTOCHOWA   
192589     2019-05-26  10:30:00  18*41'225  50*01'571                  ŻORY   
80554      2017-07-05  18:00:00  19*03'200  49*48'567         BIELSKO-BIAŁA   
172493     2019-01-24  10:00:00  19*18'534  50*34'160               MYSZKÓW   
227428     2019-12-17  15:50:00  19*04'170  50*42'423               NIERADA   
173641     2019-01-30  08:05:00  19*07'125  50*49'384           CZĘSTOCHOWA   
17023      2016-05-10  15:50:00  18*46'576  50*18'208                ZABRZE   
69164      2017-04-22  03:09:00  19*08'068  50*48'192           CZĘSTOCHOWA   
51892      2016-12-22  13:30:00  19*08'486  50*22'279                SARNÓW   
211115     2019-09-18  06:10:00  18*51'318  50*16'492           RUDA ŚLĄSKA   
27072      2016-07-14  18:55:00  19*01'572  49*50'450         BIELSKO-BIAŁA   
200991     2019-07-17  05:40:00  18*52'510  50*22'317                 BYTOM   
211764     2019-09-22  15:00:00  18*54'507  50*14'029           RUDA ŚLĄSKA   
109358     2017-12-21  22:00:00  18*55'113  50*20'465                 BYTOM   
120568     2018-03-07  10:50:00  19*02'074  49*49'509         BIELSKO-BIAŁA   
154690     2018-10-12  15:20:00  19*03'445  50*13'348              KATOWICE   
58868      2017-02-09  14:00:00  19*04'091  49*43'380            BUCZKOWICE   
131851     2018-05-19  21:30:00  18*25'025  49*56'523             GORZYCZKI   
191599     2019-05-21  05:30:00  18*30'255  50*04'356                RYBNIK   
45126      2016-11-14  21:35:00  18*49'500  50*19'071                ZABRZE   
168097     2018-12-28  16:14:00  19*00'027  50*06'034                 TYCHY   
89300      2017-09-02  14:15:00  19*06'107  50*48'193           CZĘSTOCHOWA   
37913      2016-09-30  13:55:00  18*53'555  50*08'519               MIKOŁÓW   
23617      2016-06-22  14:05:00  18*39'256  50*15'416               GLIWICE   
179440     2019-03-08  11:50:00  18*50'103  50*15'589    

### Konwersja typu danych

#### Data i czas

In [9]:
# Utworzenie kopii zbioru danych. Dalsza konwersja odbywac sie bedzie na kopii danych
df=dane.copy()

In [10]:
# Data zdarzenia i Godzina scalone do jednej kolumny
print('kolumna DATA ZDARZENIA: ', df['Data zdarzenia'].dtypes)
print('kolumna GODZINA: ', df['Data zdarzenia'].dtypes)
df['data_full'] = df['Data zdarzenia'] + ' ' + df['Godzina']
df.sample(10)

kolumna DATA ZDARZENIA:  object
kolumna GODZINA:  object


Data zdarzenia   Godzina      GPS x      GPS y    Miejscowość  \
200085     2019-07-10  19:15:00  19*13'392  49*40'209         ŻYWIEC   
173099     2019-01-28  08:25:00  18*33'078  50*05'573         RYBNIK   
170042     2019-01-09  10:00:00  18*58'219  50*06'180          TYCHY   
211216     2019-09-19  14:30:00  19*13'415  50*10'196      MYSŁOWICE   
150744     2018-09-20  09:58:00  19*25'422  50*29'038      ZAWIERCIE   
161597     2018-11-22  13:39:00  18*56'139  50*16'017        CHORZÓW   
155162     2018-10-15  11:30:00  19*15'162  50*57'180    BOGUSŁAWICE   
200509     2019-07-13  09:25:00  19*02'195  50*44'288         WYGODA   
130539     2018-05-12  08:55:00  19*02'078  49*51'026  BIELSKO-BIAŁA   
81107      2017-07-08  18:20:00  19*04'198  50*13'087       KATOWICE   

                         Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
200085               KOPERNIKA        NaN        W945             15,3   
173099                  PIASTA         35    G180296S              NaN   
170042                 BIELSKA        135      P8004S              NaN   
211216                   DŁUGA        166      P8800S              NaN   
150744  IGNACEGO PADEREWSKIEGO          5          78              120   
161597        STEFANA BATOREGO          9      P9405S              NaN   
155162               GRABOWSKA        NaN    G694001S              NaN   
200509                     NaN         28        W908              NaN   
130539               KATOWICKA        NaN      P7405S              NaN   
81107                    ADAMA          3    G100034S              NaN   

                   Obszar  ... Oznakowanie poziome Liczba pasów w zdarzeniu  \
200085  Obszar zabudowany  ...                Jest                      2.0   
173099  Obszar zabudowany  ...                Jest                      2.0   
170042  Obszar zabudowany  ...              Nie ma                      2.0   
211216  Obszar zabudowany  ...              Nie ma                      2.0   
150744  Obszar zabudowany  ...                Jest                      3.0   
161597  Obszar zabudowany  ...                Jest                      3.0   
155162  Obszar zabudowany  ...                Jest                      2.0   
200509  Obszar zabudowany  ...                Jest                      2.0   
130539  Obszar zabudowany  ...                Jest                      2.0   
81107   Obszar zabudowany  ...                Jest                      NaN   

                             Zachowanie kierującego Zachowanie pieszego  \
200085  Niezachowanie bezp. odleg. między pojazdami                 NaN   
173099  Niezachowanie bezp. odleg. między pojazdami                 NaN   
170042                               Inne przyczyny                 NaN   
211216                                          NaN                 NaN   
150744                       Nieprawidłowe omijanie                 NaN   
161597        Nieustąpienie pierwszeństwa przejazdu                 NaN   
155162                      Nieprawidłowe wymijanie                 NaN   
200509  Niedostosowanie prędkości do warunków ruchu                 NaN   
130539   Nieprzestrzeganie znaków i innych sygnałów                 NaN   
81107   Niedostosowanie prędkości do warunków ruchu                 NaN   

       Inne przyczyny Liczba rannych w zdarzeniu Liczba zabitych w zdarzeniu  \
200085            NaN                          0                           0   
173099            NaN                          0                           0   
170042            NaN                          0                           0   
211216    Stan jezdni                          0                           0   
150744            NaN                          0                           0   
161597            NaN                          0                           0   
155162            NaN                          0                           0   
200509            NaN                          0               

In [11]:
# Data zdarzenia i Godzina do typu datetime
df['data_full'] = pd.to_datetime(df['data_full'], format='%Y-%m-%d %H:%M:%S')
df.sample(10)
# TO-DO: Jaki to dzien tygodnia? Jaki to dzien miesiaca? To dzien czy noc? Jaka faza ksiezyca itd..

Data zdarzenia   Godzina      GPS x      GPS y      Miejscowość  \
81113      2017-07-08  20:40:00  19*07'447  49*36'432  WĘGIERSKA GÓRKA   
193972     2019-06-03  15:10:00  19*01'136  49*48'209    BIELSKO-BIAŁA   
47470      2016-11-28  20:10:00  18*57'050  50*16'057          CHORZÓW   
129624     2018-05-07  10:10:00  19*00'099  50*14'161         KATOWICE   
219192     2019-10-31  10:55:00  19*04'067  50*13'362         KATOWICE   
111791     2018-01-08  11:23:00  19*06'428  50*48'401      CZĘSTOCHOWA   
54848      2017-01-12  07:00:00  18*50'413  50*17'559           ZABRZE   
61880      2017-02-15  17:30:00  18*11'463  50*02'358         RACIBÓRZ   
145401     2018-08-17  14:28:00  19*08'580  50*35'536       KOZIEGŁOWY   
176714     2019-02-17  17:40:00  19*02'558  49*55'468           KANIÓW   

                           Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
81113                    ZACISZE         13    G596052S              NaN   
193972               GOLESZOWSKA         16          I2              NaN   
47470                 INWALIDZKA        NaN    G210084S              NaN   
129624                KOŚCIUSZKI        NaN      P6458S              NaN   
219192               SZOPIENICKA        NaN      P6534S              NaN   
111791            SZYMANOWSKIEGO          3    G110860S              NaN   
54848   DROGOWA TRASA ŚREDNICOWA   NIEZNANE        W902              NaN   
61880                 RACIBORSKA        NaN          45               16   
145401                       NaN   NIEZNANE           1            500,6   
176714         WINCENTEGO WITOSA         38      P4403S              NaN   

                      Obszar  ... Oznakowanie poziome  \
81113      Obszar zabudowany  ...              Nie ma   
193972     Obszar zabudowany  ...              Nie ma   
47470      Obszar zabudowany  ...                Jest   
129624     Obszar zabudowany  ...                Jest   
219192     Obszar zabudowany  ...                Jest   
111791     Obszar zabudowany  ...              Nie ma   
54848   Obszar niezabudowany  ...                Jest   
61880   Obszar niezabudowany  ...                Jest   
145401  Obszar niezabudowany  ...                Jest   
176714     Obszar zabudowany  ...                Jest   

       Liczba pasów w zdarzeniu  \
81113                       NaN   
193972                      2.0   
47470                       2.0   
129624                      4.0   
219192                      2.0   
111791                      2.0   
54848                       NaN   
61880                       NaN   
145401                      4.0   
176714                      2.0   

                                   Zachowanie kierującego Zachowanie pieszego  \
81113               Nieustąpienie pierwszeństwa przejazdu                 NaN   
193972                             Nieprawidłowe omijanie                 NaN   
47470                                      Inne przyczyny                 NaN   
129624                Nieprawidłowe zmienianie pasa ruchu                 NaN   
219192  Nieustąpienie pierwszeństwa pieszemu na przejś...                 NaN   
111791                                                NaN                 NaN   
54848                              Nieprawidłowe omijanie                 NaN   
61880                          Nieprawidłowe wyprzedzanie                 NaN   
145401        Niezachowanie bezp. odleg. między pojazdami                 NaN   
176714                            Nieprawidłowe skręcanie                 NaN   

       Inne przyczyny Liczba rannych w zdarzeniu Liczba zabitych w zdarzeniu  \
81113             NaN                          0                           0   
193972            NaN                          0                           0   
47470             NaN                          0                           0   
129624            NaN                          0                           0   
219192            NaN                          0 

#### Wspolrzedne geograficzne

Domyslny zapis wspolrzednych to D\*M'S([D]egree, [M]inute, [S]cond) np. 18\*53'289. Jednak, gdy pierwsza cyfra sekund jest zero, czasem wystepuje zapis D\*M''S np. 18\*53''11. Wystepowaly takze bledne zapisy takie jak 4\*\*56'569 i 50\*09'\*01.

In [12]:
def dmsToDecimal(coordinate_as_string):
    try:
        # Zapis domyslny to D*M'S([D]egree, [M]inute, [S]cond) zeby sprawnie podzielic kazda wspolrzedna na liste trzyelementowa
        # zamieniam ' na *, a nastepnie dziele orzymany rezultat znakami *. W trakcie konwersji danych napotkalam rozne nieprawidlowe
        # zapisy opisane w komorce powyzej. Wszystkie wystepujace nietypowe znaki takie jak **, '* i '' zamieniam na * i dziele znakiem *
        degree, minute, second = str(coordinate_as_string).replace("**", "*").replace("'*", "*").replace("''", "*").replace("'", '*').split("*")
    except:
        # Jesli nie dzieli na liste trzyelementowa to najprawdopodobniej trzeci element jest pusty(lista dwuelementowa).
        # Wtedy pierwszy element to stopien, drugi minuta. Sekundy zas sa rowne 0.
        degree, minute = str(coordinate_as_string).replace("**", "*").replace("'*", "*").replace("''", "*").replace("'", '*').split("*")
        second=0
    # W niektorych wypadkach udaje sie podzielic wsadowego stringa tak aby uzyskac ostatnia czesc jako pusty string.
    #Ponizej przeksztalcam pustego stringa w wartosc 0
    if second == "" or second == " ":
        second=0
    # Wlasciwe przeksztalcenie wspolrzednej z wartosci katowych na wartosc liczbowa
    coordinate_as_float = float(degree) + float(minute)/60 + float(second)/(60*60)
    return coordinate_as_float
# Przygotowana wyzej funkcje wywoluje na calych kolumnach dancych
df['GPS x'] = df['GPS x'].apply(lambda x: dmsToDecimal(x))
df['GPS y'] = df['GPS y'].apply(lambda x: dmsToDecimal(x))
df.sample(10)
# TO-DO Wyszukac i wykorzystac biblioteke okreslajaca wspolrzedne GPS na podstawie adresu

Data zdarzenia   Godzina      GPS x      GPS y  Miejscowość  \
139899     2018-07-12  13:20:00  19.061111  50.151111        TYCHY   
169718     2019-01-04  20:15:00  18.671667  50.378889      GLIWICE   
114477     2018-01-25  09:00:00  18.773611  50.251111       KNURÓW   
56219      2017-01-22  10:24:00  19.315278  50.244444    MYSŁOWICE   
100896     2017-11-05  01:16:00  19.217778  50.858333  CZĘSTOCHOWA   
223102     2019-11-22  10:15:00  18.678611  50.121944       RYBNIK   
126024     2018-04-12  17:30:00  19.332500  49.739167       ŻYWIEC   
106105     2017-12-04  17:25:00  19.096667  50.271111     KATOWICE   
228420     2019-12-23  16:30:00  18.992500  50.205278        TYCHY   
203644     2019-08-03  15:20:00  18.952500  50.181111      MIKOŁÓW   

                      Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
139899  JÓZEFA PIŁSUDSKIEGO         36      P8012S              NaN   
169718            KOZIELSKA        608      P7200S              NaN   
114477             DWORCOWA         15        W921              NaN   
56219                   NaN        NaN          S1            548,5   
100896          MOŚCICKIEGO         18    G110550S              NaN   
223102  STEFANA ŻEROMSKIEGO         12    G180526S              NaN   
126024             SPORYSKA        NaN      P1478S              NaN   
106105            BRYNOWSKA        NaN      P6453S              NaN   
228420         BIBLIOTECZNA         19    G190220S              NaN   
203644             KROKUSÓW         10          I2              NaN   

                      Obszar  ... Oznakowanie poziome  \
139899     Obszar zabudowany  ...                Jest   
169718     Obszar zabudowany  ...                Jest   
114477     Obszar zabudowany  ...                Jest   
56219   Obszar niezabudowany  ...                Jest   
100896     Obszar zabudowany  ...              Nie ma   
223102     Obszar zabudowany  ...              Nie ma   
126024     Obszar zabudowany  ...              Nie ma   
106105     Obszar zabudowany  ...                Jest   
228420     Obszar zabudowany  ...              Nie ma   
203644     Obszar zabudowany  ...              Nie ma   

       Liczba pasów w zdarzeniu  \
139899                      4.0   
169718                      2.0   
114477                      2.0   
56219                       NaN   
100896                      NaN   
223102                      2.0   
126024                      2.0   
106105                      NaN   
228420                      2.0   
203644                      2.0   

                                  Zachowanie kierującego Zachowanie pieszego  \
139899       Niezachowanie bezp. odleg. między pojazdami                 NaN   
169718       Niedostosowanie prędkości do warunków ruchu                 NaN   
114477       Niezachowanie bezp. odleg. między pojazdami                 NaN   
56219              Nieustąpienie pierwszeństwa przejazdu                 NaN   
100896                             Nieprawidłowe cofanie                 NaN   
223102  Nieprawidłowe wymijanie; Nieprawidłowe wymijanie                 NaN   
126024             Nieustąpienie pierwszeństwa przejazdu                 NaN   
106105       Niezachowanie bezp. odleg. między pojazdami                 NaN   
228420                             Nieprawidłowe cofanie                 NaN   
203644             Nieustąpienie pierwszeństwa przejazdu                 NaN   

       Inne przyczyny Liczba rannych w zdarzeniu Liczba zabitych w zdarzeniu  \
139899            NaN                          0                           0   
169718            NaN                          0                           0   
114477            NaN                          0                           0   
56219             NaN                          0                           0   
100896            NaN                          0                           0   
223102            NaN                          0                           0   

#### Miejsce zdarzenia



In [13]:
# Zmiana nazwy kolumn na nazwy bez spacji i polskich znakow
# df.rename(columns={'Miejscowość': 'Miejscowosc', 'Numer domu': "NrDomu", "Numer drogi": "NrDrogi"}, inplace=True)
df.columns

Index(['Data zdarzenia', 'Godzina', 'GPS x', 'GPS y', 'Miejscowość', 'Ulica',
       'Numer domu', 'Numer drogi', 'KM HM (Pikietaż)', 'Obszar',
       'Skrzyżowanie', 'Geometria drogi', 'Charakterystyka miejsca zdarzenia',
       'Oświetlenie', 'Warunki atmosferyczne', 'Rodzaj zdarzenia', 'Wina',
       'Rodzaj drogi', 'Nawierzchnia drogi', 'Stan nawierzchni',
       'Sygnalizacja świetlna', 'Oznakowanie poziome',
       'Liczba pasów w zdarzeniu', 'Zachowanie kierującego',
       'Zachowanie pieszego', 'Inne przyczyny', 'Liczba rannych w zdarzeniu',
       'Liczba zabitych w zdarzeniu', 'Liczba kolizji', 'Liczba wypadków',
       'data_full'],
      dtype='object')

In [14]:
# Pikietaz okresla miejsce zdarzenia na drodze poza terenem zabudowanym
#Zapisywany jako xx,yyy gdzie:
#xx – całkowita liczba kilometrów od początku drogi;
#yyy – całkowita liczba metrów liczona od ostatniego pełnego kilometra.
# W zbiorze danych prawdopodobnie jako wartosc w km, tyle ze z uzyciem "," jako separatora dziesietnego
def pikietazToFloat(pikietazRaw):
    if isinstance(pikietazRaw, str):
        pikietazAsFloat = float(pikietazRaw.replace(",", "."))
    else:
        pikietazAsFloat = pikietazRaw
    return pikietazAsFloat
df['KM HM (Pikietaż)']=df['KM HM (Pikietaż)'].apply(lambda x: pikietazToFloat(x))
# df.rename(columns={'KM HM (Pikietaż)': 'KM_HM_Pikietaz'})
df.sample(10)

Data zdarzenia   Godzina      GPS x      GPS y           Miejscowość  \
210121     2019-09-12  17:15:00  19.259444  50.372500             SOSNOWIEC   
201411     2019-07-19  12:01:00  19.281111  50.791944                 PORAJ   
184513     2019-04-08  12:40:00  19.224722  50.936944           CZĘSTOCHOWA   
61590      2017-02-28  05:40:00  19.096944  50.382222  SIEMIANOWICE ŚLĄSKIE   
64135      2017-03-19  18:30:00  19.056389  49.866667         BIELSKO-BIAŁA   
221318     2019-11-13  10:15:00  19.100000  50.336944              KATOWICE   
32952      2016-08-30  16:05:00  19.025278  50.386111              KATOWICE   
104950     2017-11-28  17:00:00  19.034167  50.360556  SIEMIANOWICE ŚLĄSKIE   
24291      2016-06-25  18:25:00  19.328056  50.297500              JAWORZNO   
49491      2016-12-10  09:12:00  18.863889  50.143611                SUSZEC   

                                 Ulica Numer domu Numer drogi  \
210121       STANISŁAWA MAŁACHOWSKIEGO          5    G120033S   
201411                          WESOŁA         23    G625049S   
184513                    DĄBROWSKIEGO          7      P6610S   
61590                           ŚLĄSKA         59      P9612S   
64135                     PL.RATUSZOWY          3      P7413S   
221318  TRASA NIKODEMA I JÓZEFA RENCÓW         33        W902   
32952                         SOKOLSKA        NaN      P6519S   
104950     POLSKIEGO CZERWONEGO KRZYŻA          1      P9641S   
24291                    JANA PAWŁA II        NaN          79   
49491                       PSZCZYŃSKA          1        W935   

        KM HM (Pikietaż)                Obszar  ... Oznakowanie poziome  \
210121               NaN     Obszar zabudowany  ...              Nie ma   
201411               NaN  Obszar niezabudowany  ...              Nie ma   
184513               NaN     Obszar zabudowany  ...                Jest   
61590                NaN     Obszar zabudowany  ...                Jest   
64135                NaN     Obszar zabudowany  ...              Nie ma   
221318               NaN  Obszar niezabudowany  ...                Jest   
32952                NaN     Obszar zabudowany  ...                Jest   
104950               NaN     Obszar zabudowany  ...                Jest   
24291                NaN     Obszar zabudowany  ...                Jest   
49491                NaN     Obszar zabudowany  ...                Jest   

       Liczba pasów w zdarzeniu                       Zachowanie kierującego  \
210121                      1.0                        Nieprawidłowe cofanie   
201411                      2.0                        Nieprawidłowe cofanie   
184513                      2.0                               Inne przyczyny   
61590                       2.0  Niezachowanie bezp. odleg. między pojazdami   
64135                       NaN                       Nieprawidłowe omijanie   
221318                      6.0  Niedostosowanie prędkości do warunków ruchu   
32952                       NaN                      Nieprawidłowe skręcanie   
104950                      2.0        Nieustąpienie pierwszeństwa przejazdu   
24291                       NaN  Niedostosowanie prędkości do warunków ruchu   
49491                       NaN        Nieustąpienie pierwszeństwa przejazdu   

       Zachowanie pieszego Inne przyczyny Liczba rannych w zdarzeniu  \
210121                 NaN            NaN                          0   
201411                 NaN            NaN                          0   
184513                 NaN            NaN                          0   
61590                  NaN            NaN                          0   
64135                  NaN            NaN                          0   
221318                 NaN            NaN                          0   
32952                  NaN            NaN                          0   
104950                 NaN            NaN                          0   
24291                  NaN            NaN                         

In [15]:
# Obszar: zabudowany/niezabudowany
# Przypisane zostaly wartosci logiczne: Obszar zabudowany-True, Obszar niezabudowany False.
# W razie potrzeby takie wartosci mozna wygodnie przeksztalcic na int o wartosciach odpowiednio 1 i 0
df['Obszar'].value_counts()
df['Obszar'] = df['Obszar'].apply(lambda x: True if x=='Obszar zabudowany' else False)
df.sample(10)

Data zdarzenia   Godzina      GPS x      GPS y           Miejscowość  \
216572     2019-10-17  14:30:00  19.280000  50.462222      DĄBROWA GÓRNICZA   
3194       2016-01-28  19:25:00  18.928611  50.347778           RUDA ŚLĄSKA   
9176       2016-03-05  11:00:00  18.765556  50.334722               GLIWICE   
80872      2017-07-05  15:55:00  19.037500  50.392500  SIEMIANOWICE ŚLĄSKIE   
121797     2018-03-15  08:05:00  18.956944  50.182778               MIKOŁÓW   
51200      2016-12-20  10:00:00  18.875556  50.185833               ORZESZE   
144953     2018-08-14  12:05:00  19.080000  50.390278              KATOWICE   
177450     2019-02-22  15:40:00  19.083056  50.347500              KATOWICE   
124281     2018-03-31  15:00:00  19.049167  50.334167              KATOWICE   
5034       2016-02-11  08:10:00  18.484722  50.070000                RADLIN   

                           Ulica Numer domu Numer drogi  KM HM (Pikietaż)  \
216572             ŚW. ANTONIEGO        NaN        W796               NaN   
3194                      1 MAJA        NaN        W925               NaN   
9176                         NaN       L04L         A1A               0.5   
80872                  BYTKOWSKA         19      P9602S               NaN   
121797              ŚW.WOJCIECHA         25    G380092S               NaN   
51200                  CENTRALNA        NaN          81              22.5   
144953  RONDO GEN.JERZEGO ZIĘTKA        NaN          79               NaN   
177450                KOŚCIUSZKI        243          81               NaN   
124281            KAROLA HOPPEGO         4A    G100130S               NaN   
5034                      RYMERA         55      P5028S               NaN   

        Obszar  ... Oznakowanie poziome Liczba pasów w zdarzeniu  \
216572    True  ...                Jest                      2.0   
3194      True  ...                Jest                      NaN   
9176     False  ...                Jest                      2.0   
80872     True  ...                Jest                      2.0   
121797    True  ...                Jest                      2.0   
51200    False  ...                Jest                      4.0   
144953    True  ...                Jest                      2.0   
177450    True  ...                Jest                      2.0   
124281    True  ...              Nie ma                      2.0   
5034      True  ...                Jest                      2.0   

                             Zachowanie kierującego Zachowanie pieszego  \
216572                                          NaN                 NaN   
3194          Nieudzielenie pierwszeństwa przejazdu                 NaN   
9176    Niedostosowanie prędkości do warunków ruchu                 NaN   
80872                      Nieprawidłowe Zawracanie                 NaN   
121797  Niedostosowanie prędkości do warunków ruchu                 NaN   
51200                          Zmęczenie, zaśnięcie                 NaN   
144953  Niezachowanie bezp. odleg. między pojazdami                 NaN   
177450          Nieprawidłowe zmienianie pasa ruchu                 NaN   
124281                        Nieprawidłowe cofanie                 NaN   
5034          Nieudzielenie pierwszeństwa przejazdu                 NaN   

       Inne przyczyny Liczba rannych w zdarzeniu Liczba zabitych w zdarzeniu  \
216572    Stan jezdni                          0                           0   
3194              NaN                          0                           0   
9176              NaN                          0                           0   
80872             NaN                          1                           0   
121797            NaN                          0                           0   
51200             NaN                          1                           0   
144953            NaN                          0                           0   
177450            NaN                          0                          

In [16]:
# Skrzyzowanie
# Zmiana nazwy na niezawierajaca polskich znakow
df['Skrzyżowanie'].unique()
df['Skrzyżowanie'].value_counts()
# Zamiana wartosci tekstowych na int zgodnie z zasada:
# 0 : O ruchu okrężnym; 1 : Równorzędne; 2 : Z drogą z pierwsz.
skrzyzowaniemap = {'O ruchu okrężnym':0, 'Równorzędne':1, 'Z drogą z pierwsz.':2}
df=df.applymap(lambda x: skrzyzowaniemap.get(x) if x in skrzyzowaniemap else x)
# df.rename(columns={'Skrzyżowanie': 'Skrzyzowanie'})

In [17]:
df['Geometria drogi'].value_counts()
# Geometria pozioma: 
# 0: Odcinek prosty; 1: Zakret(luk)

# Geometria pionowa: 
# 0: Spadek; 1: wzniesienie; 2: wierzcholek wzniesienia

Odcinek prosty                      142703
Zakręt, łuk                          16917
Spadek                                1866
Wzniesienie                           1317
Odcinek prosty; Spadek                 430
Odcinek prosty; Wzniesienie            329
Zakręt, łuk; Spadek                    179
Zakręt, łuk; Wzniesienie               108
Wierzchołek wzn.                        50
Odcinek prosty; Wierzchołek wzn.        16
Zakręt, łuk; Wierzchołek wzn.            5
Odcinek prosty; Zakręt, łuk              1
Name: Geometria drogi, dtype: int64

In [18]:
df['Geometria_drogi_poziom'] = df['Geometria drogi'].copy()
df['Geometria_drogi_poziom'] = df['Geometria_drogi_poziom'].apply(lambda x: 0 if 'Odcinek prosty' in str(x) else x)
df['Geometria_drogi_poziom'] = df['Geometria_drogi_poziom'].apply(lambda x: 1 if 'Zakręt' in str(x) else x)
df['Geometria_drogi_poziom'] = df['Geometria_drogi_poziom'].apply(lambda x: 1 if 'łuk' in str(x) else x)
# Co z reszta? Tzn z sytuacja kiedy nie jest zdefiniowane czy droga byla prosta czy byl to zakret?
df['Geometria_drogi_poziom'].value_counts()

0                   143479
1                    17209
Spadek                1866
Wzniesienie           1317
Wierzchołek wzn.        50
Name: Geometria_drogi_poziom, dtype: int64

In [19]:
df['Geometria_drogi_pion'] = df['Geometria drogi'].copy()
df['Geometria_drogi_pion'] = df['Geometria_drogi_pion'].apply(lambda x: 0 if 'Spadek' in str(x) else x)
df['Geometria_drogi_pion'] = df['Geometria_drogi_pion'].apply(lambda x: 1 if 'Wzniesienie' in str(x) else x)
df['Geometria_drogi_pion'] = df['Geometria_drogi_pion'].apply(lambda x: 2 if 'Wierzchołek wzn.' in str(x) else x)
# Co z reszta? Tzn z sytuacja kiedy nie jest zdefiniowane uksztaltowanie terenu?
df['Geometria_drogi_pion'].value_counts()

Odcinek prosty                 142703
Zakręt, łuk                     16917
0                                2475
1                                1754
2                                  71
Odcinek prosty; Zakręt, łuk         1
Name: Geometria_drogi_pion, dtype: int64